### PR TITLE
[feat](Nereids): persist constraint in table

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/constraint/Constraint.java [hunk 1]
```diff
@@ -17,15 +17,25 @@
...
+import java.io.DataOutput;
+import java.io.IOException;
+
+public abstract class Constraint implements Writable {
     public enum ConstraintType {
         FOREIGN_KEY("FOREIGN KEY"),
         PRIMARY_KEY("PRIMARY KEY"),
         UNIQUE("UNIQUE");
-
+        @SerializedName(value = "type_name")
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/constraint/Constraint.java:35]: nit: use short name to get smaller serailize data size

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Review by morningman — CHANGES_REQUESTED** [review]
Do not merge, I will review it later

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/constraint/TableIdentifier.java [hunk 2]
```diff
@@ -20,30 +20,41 @@
...
-        DatabaseIf databaseIf = Env.getCurrentEnv().getCurrentCatalog().getDbNullable(databaseId);
+    public TableIf toTableIf() {
+        CatalogIf<?> catalogIf = Env.getCurrentEnv().getCatalogMgr().getCatalog(catalogId);
+        if (catalogIf == null) {
+            throw new NullPointerException(String.format("Can not find catalog %s in constraint", catalogId));
+        }
+        DatabaseIf<?> databaseIf = catalogIf.getDbNullable(databaseId);
         if (databaseIf == null) {
-            throw new RuntimeException(String.format("Can not find database %s in constraint", databaseId));
+            throw new NullPointerException(String.format("Can not find database %s in constraint", databaseId));
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/constraint/TableIdentifier.java:53]: How about create a new exception? eg, MetaNotFoundException?

File: fe/fe-core/src/main/java/org/apache/doris/persist/EditLog.java [hunk 3]
```diff
@@ -973,6 +973,20 @@ public static void loadJournal(Env env, Long logId, JournalEntity journal) {
...
+                case OperationType.OP_ADD_CONSTRAINT: {
+                    final AlterConstraintLog log = (AlterConstraintLog) journal.getData();
+                    if (log != null) {
+                        log.getTableIf().replayAddConstraint(log.getConstraint());
+                    }
+                    break;
+                }
+                case OperationType.OP_DROP_CONSTRAINT: {
+                    final AlterConstraintLog log = (AlterConstraintLog) journal.getData();
+                    if (log != null) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/persist/EditLog.java:985]: Why it can be null?

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by XieJiann** [comment]
run buildall

**Review by morningman — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**